### PR TITLE
fixing compatibility issues + testsuite on compatibility

### PR DESCRIPTION
**Pull request by clebertsuconic**

### CODE REVIEW

**Comment by michaelandrepearce** [comment]
@clebertsuconic instead of this mass interface changes to pass around connection, why not add ability to simply set the packet with the channel (connection) version as it arrives.

**Comment by clebertsuconic** [comment]
@michaelandrepearce I thought it would be more complete that way.. would make sure we would have the connection during the encoding for anything else we needed.

**Comment by clebertsuconic** [comment]
@mtaylor I added hornetq to the picture here now. and it looks good.


next step is to add 1.4 as server.. and I see a few issues now.

**Comment by mtaylor** [comment]
@clebertsuconic @michaelandrepearce We actually already have ClientPacketDecoder and ServerPacketDecoder which are supposed to be used to encapsulate the decoding logic.  Why not just implement a new ServerPacketDecoder1_4.  This gets set up on connection entry so you have the version info already.  We don't use an encoder on the way out, but we could easily add something similar?

**Comment by clebertsuconic** [comment]
@mtaylor the parsing of the Message happens inside the CoreMessage.. trying to avoid copy between the incoming buffer...

Move "the hack" to PacketImpl would be a bit complex...


But i will minimize the number of changes.

**Comment by clebertsuconic** [comment]
@mtaylor the "hack" still in message but it's activated through the Packets... I still had to add connection to a lot of places.. as well as making the version-id orthogonal..

This is also fixing 1.4 as server as I added a new change.

**Comment by clebertsuconic** [comment]
@michaelandrepearce I tried to simplify as much as I could

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java [hunk 1]
```diff
@@ -752,15 +752,12 @@ public RoutingStatus route(final Message message,
...
-      if (bindings == null) {
-         // There is no queue with this address, we will check if it needs to be created
-         //         if (queueCreator.create(address)) {
-         // TODO: this is not working!!!!
-         // reassign bindings if it was created
-         //            bindings = addressManager.getBindingsForRoutingAddress(address);
-         //         }
+      if (bindings == null && context.getAddress() != null) {
+         AddressSettings settings = server.getAddressSettingsRepository().getMatch(context.getAddress().toString());
+         if (settings.isAutoCreateQueues()) {
```
Comment (clebertsuconic) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java]: @mtaylor Please review this.. this is fixing the compatibility between a 1.4 server and a 2.5 (snapshot) client.

File: artemis-core-client/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/ActiveMQSessionContext.java [hunk 2]
```diff
@@ -614,7 +614,9 @@ public void createAddress(SimpleString address,
                              Set<RoutingType> routingTypes,
                              final boolean autoCreated) throws ActiveMQException {
       CreateAddressMessage request = new CreateAddressMessage(address, routingTypes, autoCreated, true);
-      sessionChannel.sendBlocking(request, PacketImpl.NULL_RESPONSE);
+      if (!sessionChannel.getConnection().isVersionBeforeAddressChange()) {
+         sessionChannel.sendBlocking(request, PacketImpl.NULL_RESPONSE);
```
Comment (clebertsuconic) [inline artemis-core-client/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/ActiveMQSessionContext.java]: @mtaylor  this is fixing some compatibilty issues

**Comment by clebertsuconic** [comment]
@mtaylor with this last commit, this completes the full mesh. Everything compatible.. at least on these tests I have now.. I'm adding more next week.

**Comment by clebertsuconic** [comment]
DON'T MERGE THIS...

@mtaylor  help me review the auto-create stuff on that side... perhaps we only do that for previous versions... 

Testsuite is broken after this last commit..I can take out some of the auto-create, but we still need the other stuff... 


lets talk on monday...


So... again... DON'T MERGE THIS yet

**Pull request merged**